go/controller/cluster.go

```diff
@@ -50,82 +50,23 @@ func (c Cluster) GetTrainerJob(job *paddlejob.TrainingJob) (*batchv1.Job, error)
 		Get(fmt.Sprintf("%s-trainer", jobname), metav1.GetOptions{})
 }
 
-// UpdateTrainerJob updates the trainer job spec.
+// UpdateTrainerJob updates the trainer job spec
+// this will do the actual scale up/down.
 func (c Cluster) UpdateTrainerJob(job *batchv1.Job) error {
 	_, err := c.clientset.BatchV1().Jobs(job.ObjectMeta.Namespace).Update(job)
 	return err
 }
 
-func podRequestsAndLimits(pod *v1.Pod) (reqs map[v1.ResourceName]resource.Quantity, limits map[v1.ResourceName]resource.Quantity, err error) {
-	reqs, limits = map[v1.ResourceName]resource.Quantity{}, map[v1.ResourceName]resource.Quantity{}
-	for _, container := range pod.Spec.Containers {
-		for name, quantity := range container.Resources.Requests {
-			if value, ok := reqs[name]; !ok {
-				reqs[name] = *quantity.Copy()
-			} else {
-				value.Add(quantity)
-				reqs[name] = value
-			}
-		}
-		for name, quantity := range container.Resources.Limits {
-			if value, ok := limits[name]; !ok {
-				limits[name] = *quantity.Copy()
-			} else {
-				value.Add(quantity)
-				limits[name] = value
-			}
-		}
-	}
-	// init containers define the minimum of any resource
-	for _, container := range pod.Spec.InitContainers {
-		for name, quantity := range container.Resources.Requests {
-			value, ok := reqs[name]
-			if !ok {
-				reqs[name] = *quantity.Copy()
-				continue
-			}
-			if quantity.Cmp(value) > 0 {
-				reqs[name] = *quantity.Copy()
-			}
-		}
-		for name, quantity := range container.Resources.Limits {
-			value, ok := limits[name]
-			if !ok {
-				limits[name] = *quantity.Copy()
-				continue
-			}
-			if quantity.Cmp(value) > 0 {
-				limits[name] = *quantity.Copy()
-			}
-		}
-	}
-	return
-}
-
-func getPodsTotalRequestsAndLimits(podList *v1.PodList) (reqs map[v1.ResourceName]resource.Quantity, limits map[v1.ResourceName]resource.Quantity, err error) {
-	reqs, limits = map[v1.ResourceName]resource.Quantity{}, map[v1.ResourceName]resource.Quantity{}
+// getPodsTotalRequestsAndLimits accumulate resource requests and limits from all pods containers.
+func getPodsTotalRequestsAndLimits(podList *v1.PodList) (reqs v1.ResourceList, limits v1.ResourceList, err error) {
+	reqs, limits = v1.ResourceList{}, v1.ResourceList{}
 	for _, pod := range podList.Items {
-		podReqs, podLimits, err := podRequestsAndLimits(&pod)
-		if err != nil {
-			return nil, nil, err
-		}
-		for podReqName, podReqValue := range podReqs {
-			if value, ok := reqs[podReqName]; !ok {
-				reqs[podReqName] = *podReqValue.Copy()
-			} else {
-				value.Add(podReqValue)
-				reqs[podReqName] = value
-			}
-		}
-		for podLimitName, podLimitValue := range podLimits {
-			if value, ok := limits[podLimitName]; !ok {
-				limits[podLimitName] = *podLimitValue.Copy()
-			} else {
-				value.Add(podLimitValue)
-				limits[podLimitName] = value
-			}
+		for _, container := range pod.Spec.Containers {
+			AddResourceList(&reqs, container.Resources.Requests)
```

Comment:
I tried to reduce the lines by using `AddResourceList`.

Removed getting resource from `InitContainers` because currently paddle job trainer seems do not use initcontainer.

Reply:
@typhoonzero from the article we are going to use the controller in a general cluster, so could be pods other than Paddle pods.